### DISCUSSION TITLE
Issue with seed, Random and 51.4.0

### DISCUSSION THREAD

**Discussion by pkernevez** [optimatika/ojAlgo#458] [General, ojAlgo and related code]
Hi all,

In the version 51.4.0 the implementation of the class RandomNumber changed.
It now uses a instance of ThreadLocalRandom.

In our application we have test using a random Matrix.
To have a reproducible test we set the seed with the following code:
```java
    public static Primitive64Matrix getRandomNormalReturnMatrix(int timeHorizon, double assetReturn, double assetStd, int seed) {
        Normal normal = new Normal();
        normal.setSeed(seed);
        return Primitive64Matrix.FACTORY.makeFilled(timeHorizon, 1, normal)
                                        .multiply(assetStd)
                                        .add(Primitive64Matrix.FACTORY.makeFilled(timeHorizon, 1, new Deterministic(assetReturn)));
    }
```

When moving from 51.3.0 to 51.4.0 this code is now failing:
```java
java.lang.UnsupportedOperationException
	at java.base/java.util.concurrent.ThreadLocalRandom.setSeed(ThreadLocalRandom.java:200)
	at org.ojalgo.random.RandomNumber.setSeed(RandomNumber.java:98)
	at connector.slsqp4j.MatrixHelper.getRandomNormalReturnMatrix(MatrixHelper.java:17)
```
The exception is failing because the new used implementation doesn't support to set the seed:
```java
class ThreadLocalRandom 
....
    public void setSeed(long seed) {
        // only allow call from super() constructor
        if (initialized)
            throw new UnsupportedOperationException();
    }
...
```
And the new delegation in RandomNumber
https://github.com/optimatika/ojAlgo/compare/v51.3.0...v51.4.0#diff-017663c363e942692c4ab7d4c26e6b514d4a0c35dbe20480b6c82902b69d4431

```java
class RandomNumber 
...
    public void setSeed(final long seed) {
        RandomNumber.random().setSeed(seed);
    }
...
    protected final static Random random() {
        return ThreadLocalRandom.current();
    }
```

How can we set a seed with the new implementation ?

Regards,

**Comment by pkernevez** [comment]
It's simple to reproduce with this test:
```java
    @Test
    void simpleTestFailing() {
        // Given
        Normal normal = new Normal();
        // When... failed
        normal.setSeed(1L);
    }
```

**Reply by apete** [reply]
ok, I see. I'll look in to restoring the previous behaviour.

**Comment by pkernevez** [comment]
Do you need a ticket ?

**Reply by apete** [reply]
Perhaps should have been an Issue from the beginning, but no need to be so formal. Try this: https://github.com/optimatika/ojAlgo/pull/459

**Reply by pkernevez** [reply]
Seems to be good.
Thank you
